File: python/examples/trading.py
"""Place a post-only order and observe order completion separately from trade settlement.

Set AGARA_TOKEN, AGARA_TOKEN_ID and optionally AGARA_BASE_URL before running.
This example submits and cancels real orders on the configured deployment.
"""

import os

from agara_sdk import DEFAULT_BASE_URL, AgaraClient, micro_to_decimal


def main() -> None:
    with AgaraClient(
        os.environ["AGARA_TOKEN"], os.getenv("AGARA_BASE_URL", DEFAULT_BASE_URL)
    ) as client:
        accepted = client.place_order(
            token_id=os.environ["AGARA_TOKEN_ID"],
            side="BUY",
            price="0.50",
            shares="1",
            post_only=True,
        )
        try:
            order = client.wait_for_terminal(accepted["order_id"], timeout=30)
        except TimeoutError:
            client.cancel_order(accepted["order_id"])
            order = client.wait_for_terminal(accepted["order_id"], timeout=30)
        print(order["status"], order["is_terminal"])
        for trade in client.get_order_trades(accepted["order_id"])["trades"]:
            print(
                trade["status"], micro_to_decimal(trade["shares_micro"]), trade["transaction_hash"]
            )


if __name__ == "__main__":
    main()

File: rust/examples/trading.rs
//! Place a post-only limit order, cancel it, and reconcile the final state.
//!
//!     export AGARA_TOKEN="agt_..."
//!     export AGARA_TOKEN_ID="21742..."
//!     cargo run --example trading

use core::time::Duration;

use agara_sdk::{
	AgaraClient, DEFAULT_BASE_URL,
	ids::{Side, TokenId},
};

#[tokio::main]
async fn main() -> agara_sdk::Result<()> {
	let base_url = std::env::var("AGARA_BASE_URL").unwrap_or_else(|_| DEFAULT_BASE_URL.to_owned());
	let token = std::env::var("AGARA_TOKEN").expect("set AGARA_TOKEN");
	let token_id = TokenId::new(std::env::var("AGARA_TOKEN_ID").expect("set AGARA_TOKEN_ID"))?;

	let client = AgaraClient::builder().token(token.into()).base_url(base_url.into()).build()?;

	let book = client.get_orderbook(&token_id).await?;
	std::println!("best bid {:?} / ask {:?}", book.best_bid(), book.best_ask());

	let ack = client
		.place_limit_order()
		.token_id(token_id.clone())
		.side(Side::Buy)
		.price(rust_decimal_macros::dec!(0.10))
		.shares(rust_decimal_macros::dec!(1))
		.post_only(true)
		.call()
		.await?;
	std::println!("placed {} — {:?}", ack.order_id, ack.status);

	client.cancel_order(&ack.order_id).await?;
	std::println!("cancel requested");

	let order = client
		.wait_for_terminal(
			&ack.order_id,
			Duration::from_secs(10),
			Duration::from_secs(1),
		)
		.await?;
	std::println!("final status: {:?}", order.status);

	Ok(())
}
